### PR TITLE
feat: sglang DSA/sparse AIC collection + CP prefill modeling for glm5

### PR DESCRIPTION
**Pull request by YijiaZhao**
Overview:

  Add SGLang DSA data collection + AIC modeling for GLM-5-NVFP4 and DeepSeek-V4 (Pro/Flash), including context-parallel (CP) prefill modeling. Brings up the sparse-attention collectors (indexer MQA-logits, top-k, sparse FMLA) on SGLang ≥ 0.5.13 (with 0.5.12/e958 backward compatibility) and adds the CP prefill latency model to the SDK.                                       

  Details:

  Collector (SGLang DSA / sparse kernels)
  - New collector/sglang/glm5_dsa_sparse_modules.py: GLM-5 indexer sub-kernels — glm5_mqa_logits_module (deep_gemm.fp8_mqa_logits), glm5_topk_module (fast_topk, flat/top_last DELTA), glm5_dsa_attn_module (flash_mla_sparse).                                                                                                                                                        
  - collect_mla_module.py: GLM-5 DSA context/generation module collection — SGLang 0.5.13 compat (decode req.fill_len, piecewise CUDA-graph gated purely by sglang version, concat_mla.cuh regression handled), flashmla_kv backend for fp8 KV, prefix-shape validity ceilings, and removal of the per-subprocess timeout that silently truncated the isl × prefix sweep.              
  - Same-source shape derivation: sparse sub-kernel (prefix, isl, bs) shapes for BOTH context and decode are derived directly from the owning attention module's INPUT sweep (via a shared _derive_context_shapes engine) instead of being read back from the module's output perf.txt. The cheap indexer/top-k kernels therefore cover the full-chunk isl the full module drops at    
  runtime (FlashMLA scheduler-metadata smem cap), without inheriting that drop. Applies to GLM-5 (mqa/topk/dsa_attn) and DeepSeek-V4 (paged_mqa_logits/topk/csa_attn/hca_attn).
  - registry.py / registry_types.py: register the new GLM-5 sparse ops.

  SDK modeling (context parallelism)
  - operations/dsa.py: ContextDSAModule._query_cp — per-card base dsa_module(isl/cp, prefix) + MQA/top-k per-card deltas + ag_kv / ag_lse attention all-gathers (compressed-KV latent + indexer key). CP only affects attention.                                                                                                                                                       
  - models/deepseek_v32.py: pass attention_cp_size to the attention op; parallelism-balance assert now includes attention_cp_size (tp·dp·cp == moe_tp·moe_ep); shared expert FFN split by moe_tp_size (matches SGLang's shared-expert TP). MoE dispatch / shared FFN are NOT CP-scaled — CP only splits the attention token stream; the per-rank token count is computed outside the   
  ops.
  - config.py: attention_cp_size as an independent (orthogonal-to-tp_size) parallel dimension.
  - docs/CONTEXT_PARALLEL_DSA_MODELING.md: the CP round-robin token-split modeling derivation (mqa ×cp identity, top-k full/cp flat→top_last, ag_kv/ag_lse).

  Reduces to current (cp_size = 1) behavior with no change for non-CP models.

  Where should the reviewer start?

  - src/aiconfigurator/sdk/operations/dsa.py — ContextDSAModule._query_cp (the CP latency model; the core logic).
  - docs/CONTEXT_PARALLEL_DSA_MODELING.md — rationale for the per-op CP scaling, read this first.
  - src/aiconfigurator/sdk/models/deepseek_v32.py — how CP / shared-expert / MoE parallelism is wired into the op sequence (small, high-signal diff).
  - collector/sglang/glm5_dsa_sparse_modules.py + _derive_context_shapes in collector/sglang/deepseekv4_sparse_modules.py — the same-source shape derivation shared by GLM-5 and DSV4.
  - collector/sglang/collect_mla_module.py — SGLang 0.5.13 DSA collection compat (largest diff).

cp aic align for glm5, 85k - 128k, cp8 on b200(0.5.13)

Prefill: Mean MAPE vs Batch Size: Mean MAPE=5.629%, P90=16.231%

aic_vs_measured_signed_error_cases_0_prefill

bench script:
python3 ./sglang_launch_server.py \
    --model-path /workspace/cache/huggingface/hub/models--nvidia--GLM-5-NVFP4/snapshots/local \
    --trust-remote-code \
    --disable-overlap-schedule \
    --attention-backend="nsa" \
    --chunked-prefill-size=16384 \
    --disable-flashinfer-autotune \
    --enable-cache-report \
    --enable-dp-lm-head \
    --kv-cache-dtype="fp8_e4m3" \
    --max-prefill-tokens=16384 \
    --cuda-graph-max-bs 256 \
    --max-running-requests 256 \
    --mem-fraction-static=0.8 \
    --moe-dense-tp-size=1 \
    --quantization="modelopt_fp4" \
    --sampling-backend="pytorch" \
    --tp-size=8 \
    --watchdog-timeout=1000000 \
    --numa-node 0 0 0 0 1 1 1 1 \
    --enable-nsa-prefill-context-parallel \
    --attn-cp-size=8 \
    --nsa-prefill-cp-mode="round-robin-split" \
    --nsa-prefill-backend=flashmla_kv \
    --nsa-decode-backend=flashmla_kv \
    --disable-radix-cache \
    --disable-cuda-graph-padding \
    --host 0.0.0.0 \
    --port 30000

  python3 -u -m sglang.bench_serving \
    --backend sglang --dataset-name random --model <MODEL_LOCAL_PATH> \
    --random-input-len 128000 --random-output-len 10 --num-prompts 500 \
    --request-rate 50 --host 127.0.0.1 --port 30000 \
    --random-range-ratio 0.67 --disable-tqdm

## Release Notes

* **New Features**
  * Added Context Parallel (CP) support for GLM-5, including CP-aware DSA prefill latency estimation and CP-scaled MoE communication.
  * Added GLM-5 module-level sparse kernel benchmarking for **mqa logits**, **topk**, and **DSA attention** (new outputs for each module).

* **Improvements**
  * Enhanced benchmarking resilience and precision/module selection across additional edge cases, with safer eager fallbacks and updated timing/logging.
  * Updated DeepSeek-V4 sparse test generation to use static input grids.

* **Documentation**
  * Added CP DSA latency modeling documentation for GLM-5.

* **Tests**
  * Added unit tests for CP sparse-table lookup and CP DSA composition.

### CODE REVIEW

**Comment by Arsene12358** [comment]
@YijiaZhao Could you help address rabbit's comments first?

File: src/aiconfigurator/sdk/operations/dsa.py [hunk 1]
```diff
@@ -640,6 +645,171 @@ def query(self, database: PerfDatabase, **kwargs) -> PerformanceResult:
...
+            database.query_context_dsa_module(
+                b=b,
+                s=per_card,
+                prefix=prefix,
+                num_heads=self._num_heads,
+                kvcache_quant_mode=self._kvcache_quant_mode,
+                fmha_quant_mode=self._fmha_quant_mode,
+                gemm_quant_mode=self._gemm_quant_mode,
+                architecture=self._architecture,
+            )
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply src/aiconfigurator/sdk/operations/dsa.py:725]: The doc's "bf16" is compute precision, not KV-cache storage precision — these are two separate axes here.
The CP modeling doc's §4 "bf16 base" refers to the FMHA compute kernel: FlashMLA dequantizes the fp8 KV cache and runs the bf16 QkvBfloat16 kernel. In this code that axis is carried by fmha_quant_mode, which is already bf16 — not by kvcache_quant_mode.                                                                                                                         
kvcache_quant_mode is the storage precision, which is physically fp8 (fp8_ds_mla) for GLM-5-NVFP4. Forcing it to bf16 would mislabel the KV-cache byte width (2B instead of 1B) and is wrong.                                                                                                                                                                                        
The dsa_context_module data is keyed on both axes — [fmha_quant_mode][kvcache_quant_mode][gemm_quant_mode][architecture] — and the collector always collects with bf16 compute (compute_dtype is always bfloat16; the DSA/NSA FMHA kernels run bf16). So there is no "fp8-compute" row at all: the [fmha=bf16][kvcache=fp8] row already is "bf16 compute + fp8 storage" = exactly    
GLM-5's behavior. Passing the runtime kvcache_quant_mode (fp8) selects the correct row, consistent with the non-CP path.
No change needed.

File: collector/sglang/glm5_dsa_sparse_modules.py [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+    with causal [0, ke) (ke=past_kv+j+1)."""
+    from deep_gemm import fp8_mqa_logits
+
+    full_s = max(1, past_kv + M)
+    q = torch.randn(M, index_n_heads, index_head_dim, dtype=torch.bfloat16, device=device).to(torch.float8_e4m3fn)
+    k_fp8 = torch.randn(full_s, index_head_dim, dtype=torch.bfloat16, device=device).to(torch.float8_e4m3fn)
+    k_scale = torch.ones(full_s, dtype=torch.float32, device=device)
+    weights = torch.randn(M, index_n_heads, dtype=torch.float32, device=device)
+    ke = torch.arange(past_kv + 1, past_kv + M + 1, dtype=torch.int32, device=device).clamp(max=full_s)
+    ks = torch.zeros(M, dtype=torch.int32, device=device)
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply collector/sglang/glm5_dsa_sparse_modules.py]: fixed

File: docs/CONTEXT_PARALLEL_DSA_MODELING.md [hunk 3]
```diff
@@ -0,0 +1,305 @@
+# Context-Parallel (CP) Modeling for DSA Prefill
+
+How AIC estimates **one decoder layer of GLM-5 DSA (DeepSeek Sparse Attention)
+prefill under Context Parallelism** — e.g. GLM-5-NVFP4, `attn_cp_size = 8`,
+`attn_dp = 1`, `attn_tp = 1`, isl = 32768.
+
+A CP layer is modeled as **three serial parts**:
+
+```
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply docs/CONTEXT_PARALLEL_DSA_MODELING.md]: fixed

File: src/aiconfigurator/sdk/operations/dsa.py [hunk 4]
```diff
@@ -640,6 +645,171 @@ def query(self, database: PerfDatabase, **kwargs) -> PerformanceResult:
...
+        )
+        mqa_full = self._lookup_2d(g.get("mqa"), isl, prefix)
+        mqa_perc = self._lookup_2d(g.get("mqa"), per_card, prefix)
+        tl_full = self._lookup_2d(g.get("topk_last"), isl, prefix)
+        tf_perc = self._lookup_2d(g.get("topk_flat"), per_card, prefix)
+        latency = dsa_base
+        if None not in (mqa_full, mqa_perc, tl_full, tf_perc):
+            delta_mqa = mqa_full / cp - mqa_perc
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply src/aiconfigurator/sdk/operations/dsa.py:733]: fixed

**Comment by tianhaox** [comment]
recently merged task v2 entry. has this pr considered the new task definition? in yaml and default search space.

**Comment by YijiaZhao** [comment]
> recently merged task v2 entry. has this pr considered the new task definition? in yaml and default search space.
it‘s too heavy. I think we could add v2 in another pr

File: src/aiconfigurator/sdk/models/__init__.py [hunk 5]
```diff
@@ -68,6 +68,15 @@ def get_model(
...
     model_family = _architecture_to_model_family(architecture)
 
+    # Context parallelism is currently implemented only for GLM-5 DSA prefill
+    # (see ContextDSAModule._query_cp / docs/CONTEXT_PARALLEL_DSA_MODELING.md).
+    if getattr(model_config, "attention_cp_size", 1) > 1 and architecture != "GlmMoeDsaForCausalLM":
+        raise NotImplementedError(
+            f"AIC does not support context parallelism (attention_cp_size="
+            f"{model_config.attention_cp_size}) for architecture '{architecture}'. "
+            f"CP is implemented only for GLM-5 (GlmMoeDsaForCausalLM)."
+        )
```
[the remark this answers is not in the document]
Comment (Arsene12358) [reply src/aiconfigurator/sdk/models/__init__.py:78]: @coderabbitai, do we actually have any GLM-family variants that does not use GlmMoeDsaForCausalLM?

File: collector/sglang/deepseekv4_sparse_modules.py [hunk 6]
```diff
@@ -798,9 +802,163 @@ def _bench_sparse_kernel_shape(kernel, prefix, isl, bs, sc, device):
...
+        sa.cuda_graph_max_bs = None
+        sa.cuda_graph_bs = None
+        sa.tp_size = 1
+        sa.device = "cuda"
+        try:
+            sa._handle_gpu_memory_settings(get_device_memory_capacity("cuda"))
+        except Exception:
+            pass  # chunked_prefill_size is set first, before any model-dependent step
+        _CHUNKED_PREFILL = int(sa.chunked_prefill_size)
+    return _CHUNKED_PREFILL
```
[the remark this answers is not in the document]
Comment (Arsene12358) [reply collector/sglang/deepseekv4_sparse_modules.py:871]: Fixed in 122e057d — but with a `RuntimeError` guard rather than a silent fallback value.

The `TypeError` risk is real: if `_handle_gpu_memory_settings` (or `get_device_memory_capacity` itself) throws before `chunked_prefill_size` is assigned, the `except Exception: pass` swallows it and `chunked_prefill_size` stays `None`, so `int(None)` blows up.

I didn't default to `8192`, though, because this module is intentionally derive-never-default: `_dsv4_model_config` / `_dsv4_cfg_int` hard-error on missing data ("MUST NOT be guessed/defaulted"), `runtime_limits.runtime_chunk_size` already raises `RuntimeError` for this exact uninitialized-`chunked_prefill_size` condition, and this function's own docstring twice states "NOT a hardcoded 8192 and NOT a guessed ServerArgs." A silent fallback would produce a subtly-wrong shape sweep with no signal. The guard now raises a clear, actionable error instead.

`_dsv4_context_derived_shapes` only runs on a GPU box (gated by `flash_mla`/`deep_gemm` import + CUDA capability), so this path is a genuinely anomalous state, not a normal non-GPU dry-run.

**Comment by Arsene12358** [comment]
LGTM

**Review by Arsene12358 — APPROVED** [review]

File: collector/sglang/collect_dsv4_attn.py [hunk 7]
```diff
@@ -1116,12 +1116,20 @@ def kernel_func(model_runner=model_runner):
...
 
-                            with forward_context(ForwardContext(attn_backend=model_runner.attn_backend)):
+                            if _sgl.__version__.startswith("0.5.13"):
+                                from sglang.srt.model_executor.forward_context import ForwardContext, forward_context
+
+                                ctx = forward_context(ForwardContext(attn_backend=model_runner.attn_backend))
+                            else:
+                                from contextlib import nullcontext
+
+                                ctx = nullcontext()
```
[the remark this answers is not in the document]
Comment (Arsene12358) [reply collector/sglang/collect_dsv4_attn.py:1131]: Fixed in c0b0dcd5 — switched to capability detection rather than inlining the try/except.

Reused this module's existing proven helper, `collect_mla_module._import_sglang_forward_context()` (the one you pointed to), instead of duplicating the try/except: it imports `forward_context` and, on `ModuleNotFoundError`, returns a no-op shim. So every sglang that ships the module (0.5.13, 0.5.14+, …) keeps the real `ForwardContext`, and only v0.5.12/e958 — which genuinely lacks the module — falls back to the no-op. The `0.5.14+ → nullcontext` mis-selection is gone.

The helper is resolved once outside the timed `_bench_cuda_events` kernel closure (matching `collect_mla_module`'s usage at line 1308), so the import isn't repeated inside the benchmark loop.
